Implement ccip016

friedger opened the pull request:
This PR
* adds a proposal that distributes missed payouts for mia stakers as defined in CCIP-016

friedger committed:
fix: add ccip016 (WIP)

friedger committed:
fix: improve ccip 16, add simulation

whoabuddy commented on contracts/proposals/ccip016-missed-payouts.clar:
```diff
@@ -0,0 +1,421 @@
+(define-constant CCIP_016 {
+  name: "Refund Incorrect CCD007 Payouts",
+  link: "https://github.com/citycoins/governance/blob/feat/add-ccip-016/ccips/ccip-016/ccip-016-refund-incorrect-ccd007-payouts.md",
+  hash: "",
```
Noting that the [CCIP-016 governance text](https://github.com/citycoins/governance/blob/feat/add-ccip-016/ccips/ccip-016/ccip-016-refund-incorrect-ccd007-payouts.md) needs to be updated and hash can be added here.

whoabuddy commented on contracts/proposals/ccip016-missed-payouts.clar:
```diff
@@ -0,0 +1,421 @@
+    ;; check vote is complete/passed
+    (try! (is-executable))
+    ;; update vote variables
+    (var-set voteEnd block-height)
```
We could set this one to be a true 2100 block / 2 week vote since we're not rushing to have it done by a block height. [CCIP-024 did the same thing](https://explorer.hiro.so/txid/SP8A9HZ3PKST0S42VM9523Z9NV42SZ026V4K39WH.ccip024-miamicoin-signal-vote?chain=mainnet).

If we do that we'd have to switch to tenure or burn block height, so maybe better to leave as-is? Block heights move fast now!

whoabuddy commented on contracts/proposals/ccip016-missed-payouts.clar:
```diff
@@ -0,0 +1,421 @@
+(define-private (pay-all-rewards)
+  (begin
+  ;; MIA
+  (try! (contract-call? 'SP8A9HZ3PKST0S42VM9523Z9NV42SZ026V4K39WH.ccd002-treasury-mia-stacking withdraw-stx u30487 'SP32VE3A2AXWPGT7HH4B76005TJZQK7CF1MM9R0MD))
```
I think we can set both these contract names as constants, e.g. `MIA_STACKING_TREASURY` and `NYC_STACKING_TREASURY` and use that with the payout function. Same thing just a bit shorter/easier to read.

friedger replied:
This is still an open issue: https://github.com/clarity-lang/reference/issues/77

friedger replied:
Not possible today.

whoabuddy commented on contracts/proposals/ccip016-missed-payouts.clar:
```diff
@@ -0,0 +1,421 @@
+  (/ a VOTE_SCALE_FACTOR)
+)
+
+(define-private (pay-all-rewards)
```
Where are these calculations coming from? It'd be good to reference the source data in the CCIP text and resources, even better if it's easy for someone to reproduce the result.

friedger replied:
The code come from https://github.com/citycoins/scripts/pull/26

whoabuddy reviewed (changes requested):
Overall looks good! :pray: Left some feedback on what I saw in the review comments.

I would like to add some testing around the contract alongside the simulation, mostly to make sure the voting works as expected. Payout code looks clean/simple but will be important to make sure we can reproduce those numbers from the CCIP text/resources.

friedger committed:
fix: use updated script result

friedger committed:
fix: add missing stackers

friedger committed:
fix: remove unused code,

friedger committed:
chore: clarinet format

friedger committed:
chore: add hash of proposal

friedger committed:
chore: format

friedger commented:
Deployed at https://explorer.hiro.so/txid/0x79a3a2f32e163f4ceb1cd81e93808c25911a0a7fbe3b2facaa5c892c16a25045?chain=mainnet

friedger committed:
fix: use correct cycle in mainnet

whoabuddy committed:
fix: use latest docker image if possible

We're on version 3 for clarinet, probably failing check because of new epoch definition.

whoabuddy committed:
chore: bump deps to latest
